### PR TITLE
Removed panicking when frame or pipeline lookup fails.

### LINKED ISSUE #10017: Browser.html causes a panic cascade in constellation
On 075ce980a24d1e31ad704915681ef3f7edd03ccb on Linux, running `./mach run -r -w -b` gives me [this long panic](https://manishearth.pastebin.mozilla.org/8863769)

The initial panic seems to be:

```
thread 'Constellation' panicked at 'unable to find pipeline - this is a bug', ../src/libcore/option.rs:704
stack backtrace:
 
   1:     0x7f7467838f90 - sys::backtrace::tracing::imp::write::h578d6e47592a4041cHv
   2:     0x7f746783d33b - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.44995
   3:     0x7f746783cfa8 - panicking::default_handler::hffa00c845e94c6581nA
   4:     0x7f7467826a5c - sys_common::unwind::begin_unwind_inner::hd824946792143cc8vvu
   5:     0x7f7467826ee8 - sys_common::unwind::begin_unwind_fmt::h3777a045191296c5Buu
   6:     0x7f7467838401 - rust_begin_unwind
   7:     0x7f74678743df - panicking::panic_fmt::h548fe28297c9d83b7mM
   8:     0x7f746787b884 - option::expect_failed::hce0693d5039c3a72iWO
   9:     0x7f7465e8ecbc - constellation::Constellation<LTF, STF>::handle_request::h11537074669471263875
  10:     0x7f7465e73fe7 - sys_common::unwind::try::try_fn::h7432356694187898493
  11:     0x7f746783838b - __rust_try
  12:     0x7f7467834f7d - sys_common::unwind::inner_try::h8ac2089dadb4381bxsu
  13:     0x7f7465e74fda - boxed::F.FnBox<A>::call_box::h16069342354870177152
  14:     0x7f746783bde9 - sys::thread::Thread::new::thread_start::hd111df90784ef3b2llz
  15:     0x7f746380d6a9 - start_thread
  16:     0x7f746332ceec - clone
  17:                0x0 - <unknown>
```

cc @mbrubeck

### PR DESCRIPTION
**Pull request by asajeffrey**
Removed the methods `pipeline(id)`, `pipeline_mut(id)`, `frame(id)` and `frame_mut(id)` from constellation, which panicked when the table lookup failed.

The panics were causing race conditions, e.g. visiting google.com and resizing the page would cause a panic, most likely due to an iframe being added and removed, with the `DOMLoad` event arriving after the iframe had been removed, causing a panic.

This patch fixes #10017 and #8769 (although in non-webrender builds there's now a different panic, see https://github.com/servo/servo/issues/10017#issuecomment-198160200).

There are a few `TODO` items in the initial commit, for cases where it's not completely obvious what to do in the case of failure.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10082)

### CODE REVIEW

**Comment by asajeffrey** [comment]
@mbrubeck and @jdm suggest re-enabling /_mozilla/css/iframe/hide_layers2.html, since this patch removes the intermittent on headless builds.

**Comment by asajeffrey** [comment]
We might want to land #8641 before this PR.

**Comment by glennw** [comment]
This seems like a scary change.

In the original implementation I had those conditions as panics because I was fairly certain that if they do cause a panic, they signal a logic problem elsewhere.

I think that making these not panic is likely to hide / mask other genuine bugs.

Ideally I'd like to see an exact sequence of events / messages that results in these race conditions.

Thoughts @larsbergstrom  ?

**Comment by larsbergstrom** [comment]
@glennw Thanks for bringing this to my attention! I totally agree, and the reason these are scary is that when I first started working on them pipeline-related shutdown issues resulted in crashes - either the compositor or script task can get into a really horrific state (w.r.t. their native resources) if they are not shut down correclty, sometime even leading to segfaults.

I realize that the shutdown code is pretty terrifying (cleaning up our shutdown segfaults it was my first task, then glenn got to handle the intermittents, and now it appears to be one of yours <grin>). That said, now that we have session types available, I'd be totally open to instead sitting down and writing out all of the protocols and attempting to codify the shutdown sequence more explicitly.

**Comment by asajeffrey** [comment]
@glennw yup, pretty scary!

The problem is that the current code is causing panics pretty regularly (e.g. load up google.com and resize the window). The panic is caused by an unexpected sequence of events (e.g. where an iframe is loaded then removed, but its DOMLoad event arrives after the iframe's resources have been reclaimed). I think the problem is that the code requires certain sequences of events to be impossible, but the events are concurrent, so I don't see any way to enforce arrival order.

I agree that we need to rethink the protocols more thoroughly, and take a long hard look at the compositor, the question is what to do in the short term.

@larsbergstrom I'm not 100% sure session types will address the issue, but I'm certainly open to giving it a shot!

**Comment by larsbergstrom** [comment]
Maybe Thursday we can have a go at determining the new version of: shutdown

The Rust session types work was designed with exactly that diagram in mind :-) Though I'm sure things have gotten more complicated since I fixed what I could and then ran in fear in early 2014.

**Comment by asajeffrey** [comment]
@larsbergstrom In this case it's not the shutdown sequence that's causing the problem, it's an iframe being removed, and the constellation panicking because it can't send messages to the iframe. This is why I don't know whether session types will fix the problem, since it's caused by lookup failures in the pipeline hash table.

**Comment by pcwalton** [comment]
I agree with the general unease around silently ignoring frames that have unexpectedly gone missing. I do sympathize with wanting to improve browser robustness even in the presence of logic bugs though. Perhaps we could loudly print errors in these cases?

**Comment by asajeffrey** [comment]
@pcwalton we can include warning messages. These aren't exactly logic bugs though, just messages arriving in odd orders, which is pretty much par for the course on any concurrent system.

**Comment by asajeffrey** [comment]
Relevant discussion on irc: http://logs.glob.uno/?c=mozilla%23servo#c388656
Brainstorming on goals for threads in servo: https://public.etherpad-mozilla.org/p/servo-threads

**Comment by asajeffrey** [comment]
I did some digging in constellation.rs, categorizing the potential causes of panic, the results are in https://public.etherpad-mozilla.org/p/servo-threads.

TL;DR summary: there are four potential sources of panic:

1. Lookup failure in the pipeline/frame maps
2. Creating new processes
3. Creating new channels
4. Sending/receiving on those channels

This PR is just addressing (1), this still leaves a lot of (4).

**Comment by asajeffrey** [comment]
Some more irc chat, this time about what to do in the chan.send(...).unwrap() case: http://logs.glob.uno/?c=mozilla%23servo#c389192

Also @jdm points out (http://logs.glob.uno/?c=mozilla%23servo#c389143) that soldiering on in the presence of script thread failure can have nasty interactions with rooting, see #6462.

**Comment by asajeffrey** [comment]
Not entirely sure what happened there, github reckons I unassigned @glennw, no such luck :)

**Comment by drbo** [comment]
One approach is to try to use processes instead of threads whenever possible.  In that case, there should be no (or much less) need to cleanup &ndash; the OS will handle that if the process just dies.

**Comment by asajeffrey** [comment]
@drbo true, but there's still some clean-up required, e.g. when a pipeline crashes the constellation needs to update it's pipeline table, redirect the page to about:failure, etc.

**Comment by pcwalton** [comment]
Processes vs. threads is a red herring since we use unwinding when threads abnormally exit. The resource cleanup isn't the issue.

**Comment by asajeffrey** [comment]
The latest commit removes the TODO(ajeffrey) comments, and adds debug messages in each case where pipeline/frame lookup fails. 

@pcwalton: Is `debug!` enough, or do we want to complain louder than that? Note that many of the messages can fire in the case of unusual but not-actually-bugs event orders.

**Comment by asajeffrey** [comment]
I ran test-wpt locally, the PR passed except for a couple of known intermittents.

@glennw: this is now ready for review, lucky you!

**Comment by glennw** [comment]
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions.

---

*[components/compositing/constellation.rs, line 637 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVc_r00wyZ4QBbMF_A:-KDVc_r00wyZ4QBbMF_B:1315368766) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L637)):*
map -> if let

---

*[components/compositing/constellation.rs, line 892 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVgFu0_OZMo_PKiX4n:-KDVgFu0_OZMo_PKiX4o:1315368766) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L892)):*
map -> if let

---

*[components/compositing/constellation.rs, line 1007 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVgWhxIgSVPLy_f-tz:-KDVgWhxIgSVPLy_f-u-:1315368766) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1007)):*
map -> if let

---

*[components/compositing/constellation.rs, line 1122 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVgx1B_cQABJWIPlEJ:-KDVgx1B_cQABJWIPlEK:1315368766) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1122)):*
map -> if let

---

*[components/compositing/constellation.rs, line 1357 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVhco6LvCP4Sb8AtMY:-KDVhco7EbTuBzV-Ojjd:1315368766) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1357)):*
map -> if let

---

*[components/compositing/constellation.rs, line 1363 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVhxu2Jn8_izPqwlbb:-KDVhxu2Jn8_izPqwlbc:1963977497) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1363)):*
Is this needed?

---

*[components/compositing/constellation.rs, line 1451 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDViEA6M69PLHDnQXI4:-KDViEA6M69PLHDnQXI5:1301950540) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1451)):*
Should this be a match() with a debug!() message?

---

*[components/compositing/constellation.rs, line 1615 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDViO28sJIQTR82VMR2:-KDViO28sJIQTR82VMR3:1250892223) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1615)):*
Should this be a match with debug!() ?

---

*[components/compositing/constellation.rs, line 1668 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDViX3cjkV53iRnPI6s:-KDViX3cjkV53iRnPI6t:-1416101294) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1668)):*
Can this ever fail? I think in here the frames should always be available?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10082)*

**Comment by glennw** [comment]
@asajeffrey A few comments above, mostly nits.

I'm personally still not convinced this is a good idea, but it seems to do what it says on the box, so if the consensus is that this makes sense, then it mostly looks good to me.

One thing I'd suggest is doing some manual testing both with/without browser.html of navigation with the patch - the WPT tests have almost zero coverage of page navigation changes in my experience so it's very easy to cause navigation regressions without causing any WPT failures :(

**Comment by asajeffrey** [comment]
OK, code cleaned up a bit, just the issue with generating the frame tree for the compositor.

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions.

---

*[components/compositing/constellation.rs, line 1357 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVhco6LvCP4Sb8AtMY:-KDZF2ygZiqITpJgaxnH:-1870487810) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1357)):*
I think this one I'd like to keep, it turns into a nasty mess of match statements otherwise.

---

*[components/compositing/constellation.rs, line 1363 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVhxu2Jn8_izPqwlbb:-KDZFk6g9utYgW-ptX-b:65905184) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1363)):*
Good catch, the insert is already in `new_frame`.

---

*[components/compositing/constellation.rs, line 1668 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDViX3cjkV53iRnPI6s:-KDZK8vkJiVrJWRJxXxV:-177406498) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L1668)):*
I agree with you that it's probably a bug if the frame lookup fails, since a closed frame should never be a descendent of the root frame. So there should at least be some `debug!`s here. I don't think we should panic though. We could replace any missing children by `about:failure`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10082#-:-KDZMZH3nA2hgOApvzar:-1868824462)*

**Comment by asajeffrey** [comment]
@glennw for testing, my hope is to use a chaos monkey to randomly kill pipelines and make sure the constellation survives. The chaos monkey part is now in place (#10179), and I've got a PR (https://github.com/w3c/wptrunner/pull/177) in to w3c to get this called from wptrunner.

I agree re navigation regression testing, but I think that's a separate issue, I tested it manually and it seemed to work, for what that's worth.

**Comment by asajeffrey** [comment]
@glennw testing navigation, we hit https://github.com/servo/servo/issues/9984, but I think that's a separate issue. Shall I squash the commits?

**Comment by asajeffrey** [comment]
@glennw: I had to rebase anyway, so I squashed while I was at it.

**Comment by KiChjang** [comment]
@bors-servo retry

* AppVeyor being annoying

**Comment by jdm** [comment]
@bors-servo: retry
* #10300

**Comment by glennw** [comment]
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[components/compositing/constellation.rs, line 267 \[r7\]](https://reviewable.io:443/reviews/servo/servo/10082#-KEDUNnFfnHHreNXolYD:-KEDUNnG-gn9yt8Pog5m:-1285364462) ([raw file](https://github.com/servo/servo/blob/f0793916df61d8aeb621ef444c114059158286cb/components/compositing/constellation.rs#L267)):*
Should these have a debug warning? It's probably not great if the frame tree iterator can't find a frame in the current tree.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10082)*

**Comment by asajeffrey** [comment]
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[components/compositing/constellation.rs, line 637 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVc_r00wyZ4QBbMF_A:-KEDWmIa0upQpRbYlbCb:-499146009) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L637)):*
Done.

---

*[components/compositing/constellation.rs, line 892 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10082#-KDVgFu0_OZMo_PKiX4n:-KEDWld8N8h19RHxU4Qi:-499146009) ([raw file](https://github.com/servo/servo/blob/d188dd79a5b4e77cbea8575e683d48c4ad1ce997/components/compositing/constellation.rs#L892)):*
Done.

---

*[components/compositing/constellation.rs, line 267 \[r7\]](https://reviewable.io:443/reviews/servo/servo/10082#-KEDUNnFfnHHreNXolYD:-KEDXyLXvRbcDSxFcaDS:-499146009) ([raw file](https://github.com/servo/servo/blob/f0793916df61d8aeb621ef444c114059158286cb/components/compositing/constellation.rs#L267)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10082)*

**Comment by asajeffrey** [comment]
We should think about what log level to issue these messages at. At the moment it's debug, but that probably should be increased.
